az-digital/az_quickstart
# Fixes #2359 Add Curated Publication Listings.

A requested feature is for publications listings to be managed for individual people, including selectively showing them, or stickying them at the top of the listing that is provided on their person page. This should not affect ordering or display on other Person pages that may be co-authors, in other words, curation information is specific to a particular person.

## Description
This is difficult because Person cannot itself contain references to Publications _(because Publication is an optional module and Person cannot be dependent on it.)_ Information about person references likewise does not live on the Publication content type, because Publications have Authors that may or may not be associated with Person content on the site.

What's needed is a way to store context-aware curation information about the Publication listings that is not directly attached to the Person content. `DraggableViews` fills this need structurally, but this is not suited for the type of information to be stored - `DraggableViews` typically stores ordering information rather than inclusion/exclusion or sticky status.

This PR includes an **alternate field handler** for `DraggableViews`, `Quickstart Curated Views`. This alternate field handler allows users to select curation status administratively rather than using a `DraggableViews` drag element. **Curation information is stored in the context of which Person content the Publication appears** for, meaning that every Person can have their own separate curation of their publications. e.g., a Person can have a Publication stickied on their page while it is not sticky on another Person page. Person context is shared with `DraggableViews` via `hook_draggableviews_join_withargs_alter()`. 

Three Curation statuses exist:
- `Stickied` (e.g. appears at top of lists)
- `Shown` (unmodified sort order)
- `Hidden` (does not appear on that Person page)

These curation statuses are contextual, meaning that they affect only the Person Publication they are relevant to. Individual People can have individual curations.

On each person page, a contextual link appears on their Publication block, `Curate Publications`, that allows curation for that particular person.

If more complex behavior is needed, this PR could also be done with no dependency on DraggableViews for data storage, but this PR currently attempts to extend DraggableViews functionality where possible.

## Related issues
#2359 

## How to test

- Enable `az_publication`
- Create some Person content (or use the demo content Persons)
- Create publications. On the Authors you create, make sure they are linked by reference to Persons.
- Ensure that some publications are linked to multiple Persons (co-authors).
- Place the `AZ Publications: Person` block to appear on Person pages
- Visit a Person page. Verify that their publications you assigned to their matching authors are appearing
- Hover over the contextual handle of the person publication block. Select the `Curate Publications` contextual link.
- Note that the curation form mentions the Person in whose context you are curating.
- Set curation statuses for their Publications (`Stickied`, `Shown` or `Hidden`)
- Save the curation form.
- Verify the changes are reflected. Hidden publications should not be present. Stickied publications should be at the top of the list.
- Visit a co-author person of the publications. Verify that the curation statues do not affect their publications, only the person they were set on.


## Types of changes


### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
- **Minor release changes**
- **Other or unknown**

### Drupal core
- **Patch release changes**
- **Minor release changes**
- **Other or unknown**

### Drupal contrib projects
- **Patch release changes**
- **Minor release changes**
- **Other or unknown**

## Checklist
